jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java

```diff
@@ -1734,4 +1739,250 @@ public String toString() {
+                boolean isParenthesisNeeded = checkParenthesis(command, varName);
+                for (RDFNode item : items) {
+                    if (isParenthesisNeeded) {
+                        replacement.append("(");
+                    }
+                    String insert = FmtUtils.stringForNode(item.asNode(), (PrefixMapping) null);
+                    replacement.append(insert);
+                    if (isParenthesisNeeded) {
+                        replacement.append(")");
+                    }
+                    replacement.append(" ");
+                }
+                replacement.deleteCharAt(replacement.length() - 1);
+            } else {
+                replacement.append("(");
+                for (RDFNode item : items) {
+                    String insert = FmtUtils.stringForNode(item.asNode(), (PrefixMapping) null);
+                    replacement.append(insert);
+                    replacement.append(" ");
+                }
+                replacement.deleteCharAt(replacement.length() - 1);
+                replacement.append(")");
+            }
+
+            return replacement;
+        }
+
+        /**
+         * Tidy up varName if doesn't start with a ? or $.
+         *
+         * @param varName
+         * @return
+         */
+        private String createTarget() {
+            String target;
+
+            if (varName.startsWith("?") || varName.startsWith("$")) {
+                target = varName;
+            } else {
+                target = "?" + varName;
+            }
+            return target;
+        }
+
+        protected void validateValuesSafeToInject(String command, String[] targetVars) {
+
+            for (int i = 0; i < targetVars.length; i++) {
+                String targetVar = targetVars[i];
+                if (isGrouped) {
+                    //Iterate through each group according to the position of var and item.
+                    for (List<? extends RDFNode> group : groupedItems) {
+                        RDFNode item = group.get(i);
+                        validateSafeToInject(command, targetVar, item.asNode());
+                    }
+                } else {
+                    if (targetVars.length > 1) {
+                        if (items instanceof List) {
+                            //Multiple vars with items in an ordered list. Each var is checked against the item.
+                            List<? extends RDFNode> listItems = (List<? extends RDFNode>) items;
+                            RDFNode item = listItems.get(i);
+                            validateSafeToInject(command, targetVar, item.asNode());
+                        } else {
+                            //Multiple vars with items not in an ordered list. This is parsing error.
+                            throw new ARQException("Multiple VALUES variables (" + String.join(", ", targetVars) + ") being used without an ordered list of items: " + items.toString());
```

Comment:
Can this error be caught sooner at the time the user makes the original `setValues()` call?

Reply:
This error isn't relevant anymore as the `List` is being enforced by the `setValues` method. A check is performed that the number of target vars equals the number of 'items' in a 'row' with exception thrown. The check is performed late as the 'command' is not final in the class and can be changed using `setCommandText`. Could be performed earlier but the existing approach is to wait until `toString` is called.